lib/portlayer/storage/vsphere/export.go

```diff
@@ -0,0 +1,185 @@
+// Copyright 2016-2017 VMware, Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package vsphere
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"net/url"
+	"os"
+
+	"github.com/vmware/vic/lib/archive"
+	"github.com/vmware/vic/lib/portlayer/storage"
+	"github.com/vmware/vic/pkg/trace"
+	"github.com/vmware/vic/pkg/vsphere/disk"
+	"github.com/vmware/vic/pkg/vsphere/vm"
+)
+
+// Export reads the delta between child and parent volume layers, returning
+// the difference as a tar archive.
+//
+// store - the volume store containing the two layers
+// id - must inherit from ancestor if ancestor is specified
+// ancestor - the volume layer up the chain against which to diff
+// spec - describes filters on paths found in the data (include, exclude, strip)
+// data - set to true to include file data in the tar archive, false to include headers only
+func (v *VolumeStore) Export(op trace.Operation, id, ancestor string, spec *archive.FilterSpec, data bool) (io.ReadCloser, error) {
+	if ancestor != "" {
+		return nil, fmt.Errorf("volume diff is not supported in this volume store: %s", v.SelfLink.String())
+	}
+
+	l, err := v.NewDataSource(op, id)
+	if err != nil {
+		return nil, err
+	}
+
+	return l.Export(op, spec, data)
+}
+
+// NewDataSource creates and returns an DataSource associated with container storage
+func (v *VolumeStore) NewDataSource(op trace.Operation, id string) (storage.DataSource, error) {
+	uri, err := v.URL(op, id)
+	if err != nil {
+		return nil, err
+	}
+
+	source, err := v.newDataSource(op, uri)
+	if err == nil {
+		return source, err
+	}
+
+	// check for vmdk locked error here
+	if !disk.IsLockedError(err) {
+		op.Warnf("Unable to mount %s and do not know how to recover from error")
+		// continue anyway because maybe there's an online option
+	}
+
+	// online - Owners() should filter out the appliance VM
+	owners, _ := v.Owners(op, uri, disk.LockedVMDKFilter)
+	if len(owners) == 0 {
+		op.Infof("No online owners were found for %s", id)
+		return nil, errors.New("unable to create offline data source and no online owners found")
+	}
+
+	// TODO(jzt): tweak this when online export is available
```

Comment:
tweaking isn't complete - there's almost certainly extra error checking to be done, however I'm okay leaving that with @morris-jason's PR.